File: backend/utils/feature_flags.py
import os
import json
import threading
from typing import Dict, Any, Optional, Callable
import logging
from pathlib import Path
import time

# Type definition
FeatureFlag = str
FlagContext = Dict[str, Any]
FlagEvaluator = Callable[[str, FlagContext], bool]

class FeatureFlagSystem:
    """
    Enhanced feature flag system with support for environment variables,
    config files, dynamic reloading, and context-specific evaluation.
    """
    def __init__(self):
        self._lock = threading.Lock()
        self._flags: Dict[FeatureFlag, bool] = {}
        self._config_file: Optional[str] = None
        self._last_reload_time = 0
        self._reload_interval = 60  # Seconds
        self._evaluators: Dict[FeatureFlag, FlagEvaluator] = {}
        self._logger = logging.getLogger(__name__)
        
        # Load flags from environment variables during initialization
        self._load_from_environment()
    
    def _load_from_environment(self) -> None:
        """Load feature flags from environment variables with FEATURE_FLAG_ prefix"""
        with self._lock:
            env_flags = {
                key[13:]: value.lower() in ('true', '1', 'yes', 'y')
                for key, value in os.environ.items()
                if key.startswith('FEATURE_FLAG_')
            }
            self._flags.update(env_flags)
    
    def load_from_file(self, file_path: str) -> bool:
        """
        Load feature flags from a JSON config file.
        
        Args:
            file_path: Path to the JSON config file
            
        Returns:
            True if successfully loaded, False otherwise
        """
        try:
            config_path = Path(file_path)
            if not config_path.exists():
                self._logger.warning(f"Feature flag config file does not exist: {file_path}")
                return False
                
            with open(file_path, 'r') as f:
                config = json.load(f)
            
            with self._lock:
                if 'featureFlags' in config:
                    self._flags.update(config['featureFlags'])
                    self._config_file = file_path
                    self._last_reload_time = time.time()
                    return True
                else:
                    self._logger.warning("No 'featureFlags' key found in config file")
                    return False
        except Exception as e:
            self._logger.error(f"Error loading feature flags from file: {e}")
            return False
    
    def reload_if_needed(self) -> bool:
        """
        Reload flags from config file if reload interval has passed.
        
        Returns:
            True if reloaded, False otherwise
        """
        if not self._config_file:
            return False
            
        now = time.time()
        if now - self._last_reload_time > self._reload_interval:
            return self.load_from_file(self._config_file)
        return False
    
    def set_reload_interval(self, seconds: int) -> None:
        """Set the interval for automatic config file reloading"""
        if seconds < 1:
            raise ValueError("Reload interval must be at least 1 second")
        self._reload_interval = seconds
    
    def is_feature_enabled(self, flag: FeatureFlag, context: Optional[FlagContext] = None) -> bool:
        """
        Check if a feature flag is enabled, optionally considering a context.
        
        Args:
            flag: The feature flag to check
            context: Optional context dictionary for context-specific evaluation
            
        Returns:
            True if the feature is enabled, False otherwise
        """
        # Reload from file if needed
        self.reload_if_needed()
        
        # If a custom evaluator exists for this flag, use it with the context
        if context and flag in self._evaluators:
            try:
                return self._evaluators[flag](flag, context)
            except Exception as e:
                self._logger.error(f"Error in custom evaluator for flag '{flag}': {e}")
                # Fall back to static flag value
        
        # Otherwise use the static flag value
        with self._lock:
            return self._flags.get(flag, False)
    
    def set_feature_flag(self, flag: FeatureFlag, enabled: bool) -> None:
        """Set a feature flag value"""
        with self._lock:
            self._flags[flag] = enabled
    
    def register_evaluator(self, flag: FeatureFlag, evaluator: FlagEvaluator) -> None:
        """
        Register a custom evaluator function for context-based flag evaluation.
        
        Args:
            flag: The feature flag to register the evaluator for
            evaluator: Function that takes (flag, context) and returns boolean
        """
        with self._lock:
            self._evaluators[flag] = evaluator
    
    def get_all_feature_flags(self) -> Dict[FeatureFlag, bool]:
        """Get all feature flags and their current values"""
        with self._lock:
            return dict(self._flags)

# Create a singleton instance
_feature_flags = FeatureFlagSystem()

# Module-level functions for backward compatibility and ease of use
def is_feature_enabled(flag: FeatureFlag, context: Optional[FlagContext] = None) -> bool:
    """Check if a feature flag is enabled"""
    return _feature_flags.is_feature_enabled(flag, context)

def set_feature_flag(flag: FeatureFlag, enabled: bool) -> None:
    """Set a feature flag value"""
    _feature_flags.set_feature_flag(flag, enabled)

def get_all_feature_flags() -> Dict[FeatureFlag, bool]:
    """Get all feature flags and their values"""
    return _feature_flags.get_all_feature_flags()

def load_feature_flags_from_file(file_path: str) -> bool:
    """Load feature flags from a JSON config file"""
    return _feature_flags.load_from_file(file_path)

def register_context_evaluator(flag: FeatureFlag, evaluator: FlagEvaluator) -> None:
    """Register a custom context-based evaluator for a flag"""
    _feature_flags.register_evaluator(flag, evaluator)

def set_reload_interval(seconds: int) -> None:
    """Set the interval for automatic config file reloading"""
    _feature_flags.set_reload_interval(seconds) 

File: UnityClient/Assets/Scripts/Core/Utils/FeatureFlags.cs
using System;
using System.Collections.Generic;
using System.IO;
using UnityEngine;
using System.Threading;
using System.Linq;

namespace Core.Utils
{
    /// <summary>
    /// Type for feature flag identifiers (use string constants).
    /// </summary>
    public class FeatureFlag
    {
        // Common feature flags can be defined as constants here
        public const string EXAMPLE_FEATURE = "ExampleFeature";
    }
    
    /// <summary>
    /// Context for feature flag evaluation.
    /// </summary>
    [Serializable]
    public class FlagContext
    {
        public Dictionary<string, object> Values { get; set; } = new Dictionary<string, object>();
        
        public void Set<T>(string key, T value)
        {
            Values[key] = value;
        }
        
        public T Get<T>(string key, T defaultValue = default)
        {
            if (Values.TryGetValue(key, out object value) && value is T typedValue)
            {
                return typedValue;
            }
            return defaultValue;
        }
        
        public bool HasKey(string key)
        {
            return Values.ContainsKey(key);
        }
    }
    
    /// <summary>
    /// Delegate for feature flag evaluation functions.
    /// </summary>
    public delegate bool FlagEvaluator(string flag, FlagContext context);
    
    /// <summary>
    /// ScriptableObject that stores feature flag configurations and can be edited in Unity Editor.
    /// </summary>
    [CreateAssetMenu(fileName = "FeatureFlagConfig", menuName = "Config/Feature Flags")]
    public class FeatureFlagConfig : ScriptableObject
    {
        [Serializable]
        public class FlagEntry
        {
            public string Name;
            public bool Enabled;
            [TextArea(1, 5)]
            public string Description;
        }
        
        [SerializeField]
        private List<FlagEntry> flags = new List<FlagEntry>();
        
        public Dictionary<string, bool> GetFlagsDictionary()
        {
            return flags.ToDictionary(f => f.Name, f => f.Enabled);
        }
        
        /// <summary>
        /// Apply this configuration to the feature flag system.
        /// </summary>
        public void Apply()
        {
            FeatureFlagSystem.Instance.ApplyConfig(this);
        }
    }
    
    /// <summary>
    /// Thread-safe feature flag system similar to the Python implementation.
    /// </summary>
    public class FeatureFlagSystem
    {
        private static readonly object _lock = new object();
        private Dictionary<string, bool> _flags = new Dictionary<string, bool>();
        private Dictionary<string, FlagEvaluator> _evaluators = new Dictionary<string, FlagEvaluator>();
        private string _configFilePath;
        private float _lastReloadTime;
        private float _reloadInterval = 60f; // Seconds
        private bool _autoReload = false;
        
        // Singleton instance
        private static FeatureFlagSystem _instance;
        
        /// <summary>
        /// Singleton access to the FeatureFlagSystem.
        /// </summary>
        public static FeatureFlagSystem Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new FeatureFlagSystem();
                }
                return _instance;
            }
        }
        
        public FeatureFlagSystem()
        {
            LoadFromPlayerPrefs();
        }
        
        /// <summary>
        /// Applies feature flag configuration from a ScriptableObject.
        /// </summary>
        public void ApplyConfig(FeatureFlagConfig config)
        {
            if (config == null) return;
            
            lock (_lock)
            {
                Dictionary<string, bool> flagsDict = config.GetFlagsDictionary();
                foreach (var kvp in flagsDict)
                {
                    _flags[kvp.Key] = kvp.Value;
                }
            }
            
            // Save to PlayerPrefs for persistence between sessions
            SaveToPlayerPrefs();
        }
        
        /// <summary>
        /// Loads feature flags from JSON string.
        /// </summary>
        public bool LoadFromJson(string json)
        {
            try
            {
                var configObj = JsonUtility.FromJson<Wrapper<Dictionary<string, bool>>>(json);
                if (configObj != null && configObj.Value != null)
                {
                    lock (_lock)
                    {
                        foreach (var kvp in configObj.Value)
                        {
                            _flags[kvp.Key] = kvp.Value;
                        }
                    }
                    return true;
                }
                return false;
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error loading feature flags from JSON: {ex.Message}");
                return false;
            }
        }
        
        /// <summary>
        /// Loads feature flags from a JSON file.
        /// </summary>
        public bool LoadFromFile(string filePath)
        {
            if (string.IsNullOrEmpty(filePath) || !File.Exists(filePath))
            {
                Debug.LogWarning($"Feature flag config file does not exist: {filePath}");
                return false;
            }
            
            try
            {
                string json = File.ReadAllText(filePath);
                bool result = LoadFromJson(json);
                if (result)
                {
                    _configFilePath = filePath;
                    _lastReloadTime = Time.realtimeSinceStartup;
                }
                return result;
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error loading feature flags from file: {ex.Message}");
                return false;
            }
        }
        
        /// <summary>
        /// Saves current feature flags to PlayerPrefs.
        /// </summary>
        private void SaveToPlayerPrefs()
        {
            string json = JsonUtility.ToJson(new Wrapper<Dictionary<string, bool>>(_flags));
            PlayerPrefs.SetString("FeatureFlags", json);
            PlayerPrefs.Save();
        }
        
        /// <summary>
        /// Loads feature flags from PlayerPrefs.
        /// </summary>
        private void LoadFromPlayerPrefs()
        {
            if (PlayerPrefs.HasKey("FeatureFlags"))
            {
                string json = PlayerPrefs.GetString("FeatureFlags");
                LoadFromJson(json);
            }
        }
        
        /// <summary>
        /// Helper class for JSON serialization of dictionary.
        /// </summary>
        [Serializable]
        private class Wrapper<T>
        {
            public T Value;
            
            public Wrapper(T value)
            {
                Value = value;
            }
        }
        
        /// <summary>
        /// Sets automatic reload interval and enables/disables auto reload.
        /// </summary>
        public void SetReloadInterval(float seconds, bool enableAutoReload = true)
        {
            if (seconds < 1f) seconds = 1f;
            lock (_lock)
            {
                _reloadInterval = seconds;
                _autoReload = enableAutoReload;
            }
        }
        
        /// <summary>
        /// Reloads flags from config file if reload interval has passed.
        /// </summary>
        public bool ReloadIfNeeded()
        {
            if (string.IsNullOrEmpty(_configFilePath) || !_autoReload) return false;
            
            float now = Time.realtimeSinceStartup;
            if (now - _lastReloadTime > _reloadInterval)
            {
                return LoadFromFile(_configFilePath);
            }
            return false;
        }
        
        /// <summary>
        /// Checks if a feature flag is enabled, optionally considering a context.
        /// </summary>
        public bool IsFeatureEnabled(string flag, FlagContext context = null)
        {
            // Reload from file if needed
            ReloadIfNeeded();
            
            // If a custom evaluator exists for this flag, use it with the context
            if (context != null)
            {
                lock (_lock)
                {
                    if (_evaluators.TryGetValue(flag, out FlagEvaluator evaluator))
                    {
                        try
                        {
                            return evaluator(flag, context);
                        }
                        catch (Exception ex)
                        {
                            Debug.LogError($"Error in custom evaluator for flag '{flag}': {ex.Message}");
                            // Fall back to static flag value
                        }
                    }
                }
            }
            
            // Otherwise use the static flag value
            lock (_lock)
            {
                if (_flags.TryGetValue(flag, out bool enabled))
                {
                    return enabled;
                }
                return false;
            }
        }
        
        /// <summary>
        /// Sets a feature flag value.
        /// </summary>
        public void SetFeatureFlag(string flag, bool enabled)
        {
            lock (_lock)
            {
                _flags[flag] = enabled;
            }
            SaveToPlayerPrefs();
        }
        
        /// <summary>
        /// Registers a custom evaluator function for context-based flag evaluation.
        /// </summary>
        public void RegisterEvaluator(string flag, FlagEvaluator evaluator)
        {
            lock (_lock)
            {
                _evaluators[flag] = evaluator;
            }
        }
        
        /// <summary>
        /// Gets all feature flags and their current values.
        /// </summary>
        public Dictionary<string, bool> GetAllFeatureFlags()
        {
            lock (_lock)
            {
                return new Dictionary<string, bool>(_flags);
            }
        }
    }
    
    /// <summary>
    /// Static helper class for feature flags that uses the singleton FeatureFlagSystem.
    /// </summary>
    public static class FeatureFlagHelper
    {
        /// <summary>
        /// Checks if a feature flag is enabled.
        /// </summary>
        public static bool IsFeatureEnabled(string flag, FlagContext context = null)
        {
            return FeatureFlagSystem.Instance.IsFeatureEnabled(flag, context);
        }
        
        /// <summary>
        /// Sets a feature flag value.
        /// </summary>
        public static void SetFeatureFlag(string flag, bool enabled)
        {
            FeatureFlagSystem.Instance.SetFeatureFlag(flag, enabled);
        }
        
        /// <summary>
        /// Gets all feature flags and their values.
        /// </summary>
        public static Dictionary<string, bool> GetAllFeatureFlags()
        {
            return FeatureFlagSystem.Instance.GetAllFeatureFlags();
        }
        
        /// <summary>
        /// Loads feature flags from a JSON file.
        /// </summary>
        public static bool LoadFromFile(string filePath)
        {
            return FeatureFlagSystem.Instance.LoadFromFile(filePath);
        }
        
        /// <summary>
        /// Registers a custom context-based evaluator for a flag.
        /// </summary>
        public static void RegisterEvaluator(string flag, FlagEvaluator evaluator)
        {
            FeatureFlagSystem.Instance.RegisterEvaluator(flag, evaluator);
        }
        
        /// <summary>
        /// Sets automatic file reload interval and enables/disables auto reload.
        /// </summary>
        public static void SetReloadInterval(float seconds, bool enableAutoReload = true)
        {
            FeatureFlagSystem.Instance.SetReloadInterval(seconds, enableAutoReload);
        }
    }
    
#if UNITY_EDITOR
    /// <summary>
    /// Custom editor for FeatureFlagConfig to provide a better editing experience.
    /// </summary>
    [UnityEditor.CustomEditor(typeof(FeatureFlagConfig))]
    public class FeatureFlagConfigEditor : UnityEditor.Editor
    {
        public override void OnInspectorGUI()
        {
            base.OnInspectorGUI();
            
            FeatureFlagConfig config = (FeatureFlagConfig)target;
            
            UnityEditor.EditorGUILayout.Space();
            if (GUILayout.Button("Apply Configuration"))
            {
                config.Apply();
                Debug.Log("Feature flag configuration applied to runtime system");
            }
            
            if (GUILayout.Button("Export to JSON"))
            {
                string path = UnityEditor.EditorUtility.SaveFilePanel(
                    "Export Feature Flags",
                    Application.dataPath,
                    "featureFlags.json",
                    "json");
                    
                if (!string.IsNullOrEmpty(path))
                {
                    try
                    {
                        var wrapper = new { featureFlags = config.GetFlagsDictionary() };
                        string json = JsonUtility.ToJson(wrapper, true);
                        File.WriteAllText(path, json);
                        Debug.Log($"Feature flags exported to {path}");
                    }
                    catch (Exception ex)
                    {
                        Debug.LogError($"Error exporting feature flags: {ex.Message}");
                    }
                }
            }
        }
    }
#endif
} 